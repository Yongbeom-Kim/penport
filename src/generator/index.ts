export * from "./generator";
export * from "./tailwind_generator";