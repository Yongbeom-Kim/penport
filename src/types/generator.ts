import { PathLike } from "node:fs";
import { Color } from "./penpot";

import { Typography } from "./penpot";
import { AssertionError } from "./errors";

export interface OutputFile {
  path: PathLike;
  contents: string;
}

export type GeneratorOptions = {
  cssOutputPath: PathLike;
  outputMode: "pure-css";
  typography: {
    fallbackToSansSerif: boolean; // default: true
  }
}

export type Generator = (data: {
  typographies: Typography[],
  colors: Color[]
}, options: GeneratorOptions) => OutputFile[];

export class CSSDeclaration {
  constructor(
    private _name: string,
    private _value: string | number,
    public readonly comment?: string,
  ) {
    if (this._name.includes(" ")) {
      throw new AssertionError(`CSS declaration name cannot contain spaces, name: ${this._name}`);
    }
    if (this._name.includes(".")) {
      throw new AssertionError(`CSS declaration name cannot contain dots, name: ${this._name}`);
    }
    if (this._name !== this._name.toLowerCase()) {
      throw new AssertionError(`CSS declaration name must be lowercase, name: ${this._name}`);
    }
  }

  toString() {
    return `${this.name}: ${this.value}; ${this.comment ? `/* ${this.comment} */` : ""}`;
  }

  get name() {
    return this._name;
  }

  get value() {
    return this._value;
  }
}

export class EmptyCSSDeclaration extends CSSDeclaration {
  constructor(comment?: string) {
    super("", "", comment);
  }

  override toString() {
    if (this.comment) {
      return `/* ${this.comment} */`;
    }
    return "";
  }
}

export class CSSVariable extends CSSDeclaration {
  constructor(name: string, value: string | number) {
    super(`--${name}`, value);
  }
}

export class CSSDeclarationBlock {
  constructor(
    public readonly selector: string,
    public readonly declarations: CSSDeclaration[],
  ) {}

  static merge(...blocks: CSSDeclarationBlock[]): CSSDeclarationBlock[] {
    if (blocks.length === 0) {
      throw new AssertionError("Cannot merge empty blocks");
    }
    const blocksBySelector: Record<string, CSSDeclarationBlock[]> = {};
    for (const block of blocks) {
      blocksBySelector[block.selector] = [...(blocksBySelector[block.selector] || []), block];
    }

    return Object.values(blocksBySelector).map(blocks => this._mergeSameSelector(...blocks));
  }

  static _mergeSameSelector(...blocks: CSSDeclarationBlock[]): CSSDeclarationBlock {
    if (blocks.length === 0) {
      throw new AssertionError("Cannot merge empty blocks");
    }
    if ((new Set(blocks.map(block => block.selector))).size !== 1) {
      throw new AssertionError(`Cannot merge blocks with different selectors. Selectors: ${blocks.map(block => block.selector).join(", ")}`);
    }

    const finalBlock = new CSSDeclarationBlock(blocks[0]!.selector, []);

    for (const block of blocks) {
      if (finalBlock.declarations.length !== 0) {
        finalBlock.declarations.push(new EmptyCSSDeclaration())
      }
      finalBlock.declarations.push(...block.declarations);
    }

    return finalBlock;
  }

  toString() {
    return this.selector + " {\n" +
      "  " + this.declarations.map((declaration) => declaration.toString()).join("\n  ") + "\n" +
      "}";
  }
}

export class CSSClass extends CSSDeclarationBlock {
  constructor(
    className: string,
    declarations: CSSDeclaration[],
  ) {
    super(`.${className}`, declarations);
    if (className.includes(" ")) {
      throw new AssertionError(`CSS class name cannot contain spaces, name: ${className}`);
    }
    if (className.includes(".")) {
      throw new AssertionError(`CSS class name cannot contain dots, name: ${className}`);
    }
  }
}

export class RootCSSDeclarationBlock extends CSSDeclarationBlock {
  constructor(
    declarations: CSSDeclaration[],
  ) {
    super(":root", declarations);
  }
}