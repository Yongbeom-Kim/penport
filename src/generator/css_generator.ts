import { AssertionError } from "../types/errors";
import { CSSClass, CSSDeclaration, CSSDeclarationBlock, CSSVariable, EmptyCSSDeclaration, PureCSSGeneratorOptions, OutputFile, RootCSSDeclarationBlock } from "../types/generator";
import { Color, Typography } from "../types/penpot";

export const pureCssGenerator = (theme: { typographies: Typography[], colors: Color[] }, options: PureCSSGeneratorOptions): OutputFile[] => {
  const { typographies, colors } = theme;
  const files: OutputFile[] = [];
  const typographyBlocks = generateCssTypography(typographies, options);
  const colorBlocks = generateCssColors(colors);

  const mergedBlocks = CSSDeclarationBlock.merge(...colorBlocks, ...typographyBlocks);

  files.push({
    path: options?.cssOutputPath ?? "styles.css",
    contents: mergedBlocks.map(block => block.toString()).join("\n\n")
  });

  return files;
};

const getCssName = (...fullPath: string[]) => {
  if (fullPath.length === 0) {
    throw new AssertionError("Cannot get CSS name, fullPath is required and cannot be empty");
  }
  return fullPath
    .filter(path => path !== undefined && path !== null && path !== "")
    .join("_")
    .toLowerCase()
    .replaceAll("/", "_")
    .replaceAll(" ", "-")
    .replaceAll(".", "_")
    .replaceAll(/[^a-z0-9_-]/g, "");
}

const generateCssTypography = (typographies: Typography[], options: PureCSSGeneratorOptions): CSSDeclarationBlock[] => {
  const uniqueFontFamilies = [...new Set(typographies.map((typography) => typography.fontFamily))];

  const cssBlocks: CSSDeclarationBlock[] = [];

  const fontFamilyCssVariables: Record<string, CSSVariable> = {};

  uniqueFontFamilies.forEach((fontFamily) => {
    const cssVariableName = `font-family-${fontFamily.trim().replaceAll(/\s+/g, "-").toLowerCase()}`;
    if (options.typography.fallbackToSansSerif) {
      fontFamilyCssVariables[fontFamily] = new CSSVariable(cssVariableName, `"${fontFamily}", sans-serif`);
    } else {
      fontFamilyCssVariables[fontFamily] = new CSSVariable(cssVariableName, `"${fontFamily}"`);
    }
  });

  cssBlocks.push(new RootCSSDeclarationBlock([
    new EmptyCSSDeclaration("Add your own fallback fonts here:"),
    ...Object.values(fontFamilyCssVariables),
  ]));

  for (const typography of typographies) {
    const className = getCssName(typography.path ?? "", typography.name);
    
    cssBlocks.push(new CSSClass(className, [
      new CSSDeclaration("font-family", `var(${fontFamilyCssVariables[typography.fontFamily]!.name})`),
      new CSSDeclaration("font-size", `${parseFloat((typography.fontSize / 16).toFixed(4))}rem`),
      new CSSDeclaration("font-style", typography.fontStyle),
      new CSSDeclaration("font-weight", typography.fontWeight),
      new CSSDeclaration("line-height", typography.lineHeight.toString()),
      new CSSDeclaration("letter-spacing", `${parseFloat((typography.letterSpacing / typography.fontSize).toFixed(4))}em`),
      new CSSDeclaration("text-transform", typography.textTransform)
    ]));
  }

  return cssBlocks;
}

const generateCssColors = (colors: Color[]): CSSDeclarationBlock[] => {
  const cssBlocks: CSSDeclarationBlock[] = [];

  cssBlocks.push(new RootCSSDeclarationBlock(
    colors.map((color) => {
      let cssVariableName = getCssName(color.path ?? "", color.name);
      if (cssVariableName.length === 0) {
        throw new AssertionError(`Cannot generate CSS variable name, color: ${color}`);
      }
      if (!(/^[a-z]/.test(cssVariableName))) {
        cssVariableName = `color-${cssVariableName}`;
      }
      return new CSSVariable(cssVariableName, color.color);
    })
  ));

  return cssBlocks;
}