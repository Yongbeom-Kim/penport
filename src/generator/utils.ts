import { AssertionError } from "../types/errors";
import { Color } from "../types/penpot";

export const sanitizeCssIdentifier = (...fullPath: string[]) => {
  if (fullPath.length === 0) {
    throw new AssertionError("Cannot get CSS name, fullPath is required and cannot be empty");
  }
  return fullPath
    .filter(path => path !== undefined && path !== null && path !== "")
    .join("_")
    .toLowerCase()
    .replaceAll("/", "_")
    .replaceAll(" ", "-")
    .replaceAll(".", "_")
    .replaceAll(/[^a-z0-9_-]/g, "");
}

export const getColorCssName = (color: Color): string => {
  let cssVariableName = sanitizeCssIdentifier(color.path ?? "", color.name);
  
  if (cssVariableName === "") {
    throw new AssertionError(`Cannot generate CSS variable name for color: ${color.name}`);
  }
  
  if (!/^[a-z]/.test(cssVariableName)) {
    cssVariableName = `color-${cssVariableName}`;
  }
  
  return cssVariableName;
};

export const getColorValue = (color: Color): string => {
  if (color.opacity === 1) {
    return color.color;
  } else {
    const rgb = hexToRgb(color.color);
    return `rgba(${rgb}, ${color.opacity})`;
  }
};

export const hexToRgb = (hex: string): string => {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  return `${r}, ${g}, ${b}`;
};

export const pxToRem = (px: number): number => {
  return px / 16;
};

export const mapFontWeight = (weight: string): string => {
  const weightMap: Record<string, string> = {
    "100": "font-thin",
    "200": "font-extralight", 
    "300": "font-light",
    "400": "font-normal",
    "500": "font-medium",
    "600": "font-semibold",
    "700": "font-bold",
    "800": "font-extrabold",
    "900": "font-black"
  };
  
  return weightMap[weight] || "font-normal";
};

export const mapFontFamily = (family: string): string => {
  const sansSerifFonts = ["Inter", "Arial", "Helvetica", "sans-serif"];
  const serifFonts = ["Times", "Georgia", "serif"];
  const monoFonts = ["Monaco", "Consolas", "monospace"];
  
  if (sansSerifFonts.some(font => family.toLowerCase().includes(font.toLowerCase()))) {
    return "font-sans";
  }
  if (serifFonts.some(font => family.toLowerCase().includes(font.toLowerCase()))) {
    return "font-serif";
  }
  if (monoFonts.some(font => family.toLowerCase().includes(font.toLowerCase()))) {
    return "font-mono";
  }
  
  return "font-sans";
};

export const getTypographyCssName = (path: string, name: string): string => {
  return sanitizeCssIdentifier(path ?? "", name);
};

export const sanitizeFontFamilyName = (fontFamily: string): string => {
  return sanitizeCssIdentifier(fontFamily);
};

export const formatLetterSpacing = (spacing: number): string => {
  return spacing.toFixed(4).replace(/\.?0+$/, "");
};

export const mapTextTransform = (textTransform: string): string => {
  const transformMap: Record<string, string> = {
    none: "",
    capitalize: "capitalize",
    uppercase: "uppercase",
    lowercase: "lowercase",
  };

  return transformMap[textTransform] || "";
};

export const getFontFamilyCssVariableName = (fontFamily: string): string => {
  return `font-family-${fontFamily.trim().replaceAll(/\s+/g, "-").toLowerCase()}`;
};