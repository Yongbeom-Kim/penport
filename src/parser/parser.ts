import { PathLike } from "node:fs";
import { isColorEntry, isTypographyEntry, unzipPenpotFile } from "./file_parser";
import { parseColorData, parseTypographyData } from "./data_parser";
import { Color, Typography } from "../types/penpot";

export const parsePenpotFile = (penpotFile: PathLike | Buffer): { typographies: Typography[], colors: Color[] } => {
  const unzippedPenpotFile = unzipPenpotFile(penpotFile);
  const entries = unzippedPenpotFile.getEntries();
  
  const typographies = entries
    .filter(isTypographyEntry)
    .map(entry => entry.getData().toString())
    .map(data => parseTypographyData(data));
  const colors = entries
    .filter(isColorEntry)
    .map(entry => entry.getData().toString())
    .map(data => parseColorData(data));

  return { typographies, colors };
}