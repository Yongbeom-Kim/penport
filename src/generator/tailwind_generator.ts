import { AssertionError } from "../types/errors";
import { OutputFile, TailwindGeneratorOptions } from "../types/generator";
import { Color, Typography } from "../types/penpot";


export const tailwindGenerator = (theme: {
  typographies: Typography[],
  colors: Color[]
}, options: TailwindGeneratorOptions): OutputFile[] => {
  if (!options.configOutputPath.toString().endsWith(".js")) {
    throw new AssertionError(`Config output path must end with .js, path: ${options.configOutputPath}`);
  }

  if (!options.cssOutputPath.toString().endsWith(".css")) {
    throw new AssertionError(`CSS output path must end with .css, path: ${options.cssOutputPath}`);
  }

  const outputs: OutputFile[] = [];

  // Generate Tailwind config file
  const configContent = generateTailwindConfig(theme.colors);
  outputs.push({
    path: options.configOutputPath,
    contents: configContent
  });

  // Generate typography CSS file
  const cssContent = generateTypographyCSS(theme.typographies);
  outputs.push({
    path: options.cssOutputPath,
    contents: cssContent
  });

  return outputs;
};

function generateTailwindConfig(colors: Color[]): string {
  const colorEntries = colors.map(color => {
    const colorValue = color.opacity === 1 
      ? `"${color.color}"`
      : `"rgba(${hexToRgb(color.color)}, ${color.opacity})"`;
    
    return `    "${color.name}": ${colorValue}`;
  });

  return `module.exports = {
  theme: {
    extend: {
      colors: {
${colorEntries.join(',\n')}
      }
    }
  }
};`;
}

function generateTypographyCSS(typographies: Typography[]): string {
  const utilityClasses = typographies.map(typography => {
    const fontSize = pxToRem(typography.fontSize);
    const lineHeight = pxToRem(typography.lineHeight);
    const fontWeight = mapFontWeight(typography.fontWeight);
    const letterSpacing = `${typography.letterSpacing / typography.fontSize}em`;
    const fontFamily = mapFontFamily(typography.fontFamily);

    return `.text-${typography.name} {
  @apply text-[${fontSize}rem] leading-[${lineHeight}rem] ${fontWeight} tracking-[${letterSpacing}] ${fontFamily};
}`;
  });

  return `@tailwind base;
@tailwind components;
@tailwind utilities;

${utilityClasses.join('\n\n')}`;
}

function hexToRgb(hex: string): string {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  return `${r}, ${g}, ${b}`;
}

function pxToRem(px: number): number {
  return px / 16;
}

function mapFontWeight(weight: string): string {
  const weightMap: Record<string, string> = {
    "100": "font-thin",
    "200": "font-extralight", 
    "300": "font-light",
    "400": "font-normal",
    "500": "font-medium",
    "600": "font-semibold",
    "700": "font-bold",
    "800": "font-extrabold",
    "900": "font-black"
  };
  
  return weightMap[weight] || "font-normal";
}

function mapFontFamily(family: string): string {
  // Default to font-sans for common sans-serif fonts
  const sansSerifFonts = ["Inter", "Arial", "Helvetica", "sans-serif"];
  const serifFonts = ["Times", "Georgia", "serif"];
  const monoFonts = ["Monaco", "Consolas", "monospace"];
  
  if (sansSerifFonts.some(font => family.toLowerCase().includes(font.toLowerCase()))) {
    return "font-sans";
  }
  if (serifFonts.some(font => family.toLowerCase().includes(font.toLowerCase()))) {
    return "font-serif";
  }
  if (monoFonts.some(font => family.toLowerCase().includes(font.toLowerCase()))) {
    return "font-mono";
  }
  
  return "font-sans"; // Default fallback
}