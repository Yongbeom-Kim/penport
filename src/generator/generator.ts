import { AssertionError } from "../types/errors";
import { GeneratorOptions } from "../types/generator";
import { Color } from "../types/penpot";
import { Typography } from "../types/penpot";
import { deepMerge, DeepPartial } from "../types/utils";
import { pureCssGenerator } from "./css_generator";


const defaultGeneratorOptions: GeneratorOptions = {
  cssOutputPath: "styles.css",
  outputMode: "pure-css",
  typography: {
    fallbackToSansSerif: true,
  },
}

export const generator = (theme: {
  typographies: Typography[],
  colors: Color[]
}, options: DeepPartial<GeneratorOptions>) => {
  const completeOptions = deepMerge(defaultGeneratorOptions, options ?? {});

  if (!completeOptions.cssOutputPath.toString().endsWith(".css")) {
    throw new AssertionError(`CSS output path must end with .css, path: ${completeOptions.cssOutputPath}`);
  }

  switch (completeOptions.outputMode) {
    case "pure-css":
      return pureCssGenerator(theme, completeOptions);
    default:
      throw new AssertionError(`Unsupported output mode: ${completeOptions.outputMode}`);
  }
}