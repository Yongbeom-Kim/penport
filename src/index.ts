#!/usr/bin/env node
import { program } from "commander";
import { writeFileSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import { initCommand } from './scripts/init';
import { validateEnvironment, getAccessToken, getFileId } from './utils/config';
import { getFile } from './api/request';
import { generator } from './generator/generator';

program
  .command("init")
  .description("Initialize penport configuration")
  .action(initCommand);

program
  .command("export")
  .description("Export all themes from Penpot")
  .option("--format <format>", "Output format: css or tailwind", "css")
  .option("--config-output <path>", "Output path for Tailwind config (tailwind format only)", "penpot.tailwind.config.js")
  .option("--css-output <path>", "Output path for CSS file", "styles.css")
  .action(async (options) => {
    try {
      // Validate configuration files exist
      console.log("Validating configuration...");
      validateEnvironment();
      
      // Fetch data from Penpot API
      console.log("Fetching themes from Penpot...");
      const penpotFile = await getFile(getFileId(), getAccessToken());
      
      // Transform API response to theme format
      const theme = {
        typographies: Object.values(penpotFile.data.typographies),
        colors: Object.values(penpotFile.data.colors)
      };
      
      console.log(`Found ${theme.typographies.length} typographies and ${theme.colors.length} colors`);
      
      // Generate output files
      console.log(`Generating ${options.format} output...`);
      const generatorOptions = options.format === "tailwind" 
        ? {
            outputMode: "tailwind" as const,
            configOutputPath: options.configOutput,
            cssOutputPath: options.cssOutput
          }
        : {
            outputMode: "pure-css" as const,
            cssOutputPath: options.cssOutput
          };
      
      const outputFiles = generator(theme, generatorOptions);
      
      // Write files to disk
      for (const file of outputFiles) {
        const dirPath = dirname(file.path.toString());
        mkdirSync(dirPath, { recursive: true });
        writeFileSync(file.path.toString(), file.contents, 'utf-8');
        console.log(`Generated: ${file.path}`);
      }
      
      console.log("Export completed successfully!");
      
    } catch (error) {
      console.error("Export failed:", error instanceof Error ? error.message : error);
      process.exit(1);
    }
  });

program.parse(process.argv);
