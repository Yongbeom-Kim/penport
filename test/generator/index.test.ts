import { generator } from "../../src/generator/generator";
import { parsePenpotFile } from "../../src/parser";

describe("parse Base Test Case penpot file", () => {
  it("should unzip a penpot file", () => {
    const filePath = "test/resources/Base Test Case.penpot";
    const cssOutputPath = "src/output.css";
    const parsedFile = parsePenpotFile(filePath);
    const generatedFiles = generator(parsedFile, { outputMode: "pure-css", cssOutputPath });
    expect(generatedFiles.length).toBe(1);
    expect(generatedFiles[0]!.path.toString()).toBe(cssOutputPath);
    expect(generatedFiles).toMatchSnapshot();
  });
});