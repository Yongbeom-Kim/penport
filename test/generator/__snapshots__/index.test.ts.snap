// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse Base Test Case penpot file should unzip a penpot file 1`] = `
[
  {
    "contents": ":root {
  --cy_tailwind-cyan-50: #ecfeff; 
  --id_tailwind-indigo-500-a-50: #6366f1; 
  --color-1e2932: #1e2932; 
  
  /* Add your own fallback fonts here: */
  --font-family-source-sans-3: "Source Sans 3", sans-serif; 
  --font-family-stalemate: "Stalemate", sans-serif; 
}

.ssp_source-sans-3-14-regular {
  font-family: var(--font-family-source-sans-3); 
  font-size: 0.875rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 1.2; 
  letter-spacing: 0em; 
  text-transform: none; 
}

.ssp_source-sans-3-bolditalic {
  font-family: var(--font-family-source-sans-3); 
  font-size: 1.125rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 1.2; 
  letter-spacing: 0em; 
  text-transform: none; 
}

.ssp_source-sans-3-200 {
  font-family: var(--font-family-source-sans-3); 
  font-size: 0.625rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 1.2; 
  letter-spacing: 0em; 
  text-transform: none; 
}

.st_stalemate-regular {
  font-family: var(--font-family-stalemate); 
  font-size: 0.625rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 1.2; 
  letter-spacing: 0em; 
  text-transform: capitalize; 
}

.st_stalemate-regular-24-line-spacing {
  font-family: var(--font-family-stalemate); 
  font-size: 1.5rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 1.2; 
  letter-spacing: 0em; 
  text-transform: none; 
}

.st_stalemate-regular-all-caps-24 {
  font-family: var(--font-family-stalemate); 
  font-size: 2.6875rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 3; 
  letter-spacing: 0.0558em; 
  text-transform: uppercase; 
}

.no-group {
  font-family: var(--font-family-source-sans-3); 
  font-size: 0.875rem; 
  font-style: normal; 
  font-weight: 400; 
  line-height: 1.2; 
  letter-spacing: 0em; 
  text-transform: none; 
}",
    "path": "src/output.css",
  },
]
`;
